drivers/syslog/syslog_filechannel.c

```diff
@@ -108,6 +168,12 @@ FAR struct syslog_channel_s *syslog_file_channel(FAR const char *devpath)
       syslog_dev_uninitialize(g_syslog_file_channel);
     }
 
+  /* Rotate the log file, if needed. */
+
+#ifdef CONFIG_SYSLOG_FILE_ROTATE
+  log_rotate(devpath);
```

Comment:
how to handle the long live device? some device may run many day and then the log size will become very very big. It's better to detect the file size reach to the limitation in the write/putc/force_putc callback. We can save the initial file size after opening the file, and increase this value in these callback and rename the file once the limitation reach.

Reply:
Theoretically you are right. This can happen.  
However, in practice this is very rare, and I prefer to keep the simple solution.

What you describe is a good idea, but I am afraid that it won't work easily/correctly.  
There is no easy way to manipulate the log after it is opened without side-effects:
* Either the system will block during the rotation, possibly harming high priority threads.
* Or there will be lost logs.

Unfortunately, it is not possible to predict when the rotation will take place. This will have an impact on the real-time behavior of the system. Keeping the rotations during opening only keeps the system predictable.

Reply:
> Theoretically you are right. This can happen.
> However, in practice this is very rare, and I prefer to keep the simple solution.
> 

It isn't rare, many IoT product will run a long time.

> What you describe is a good idea, but I am afraid that it won't work easily/correctly.
> There is no easy way to manipulate the log after it is opened without side-effects:

The action is simple:

1. Check the total size after writing log
2. close/rename/open file if the limitation reach

> 
> * Either the system will block during the rotation, possibly harming high priority threads.
> * Or there will be lost logs.
> 

The rename process for all filesystem implementaiton should be fast and not relate to the file size.

> Unfortunately, it is not possible to predict when the rotation will take place. This will have an impact on the real-time behavior of the system. Keeping the rotations during opening only keeps the system predictable.

Normally, write data to file isn't simpler than rename a file. So if renaming impact the real-time behaviour very much, write data should have the same impact.

If the real-time is a concern, the right approach is:

1. Send the log to ramlog
2. log service read ramlog and write to file

Reply:
> It isn't rare, many IoT product will run a long time.

I mean, it is rare to have a system run for so long AND produce so much *system* logs at the same time that fill up today's huge memories. Most IoT devices are low-power, and do not have very verbose logs...

To be honest, I don't ever expect a several GB SD to fill-up. My primary concern is to somehow constrain the log size, in case I need to upload it to an external server. So it doesn't get into gigabyte scale after several months/years.  
This is why I don't emphasize on log size accuracy, rather than performance.

> The rename process for all filesystem implementaiton should be fast and not relate to the file size.

I am not that much concerned about how fast the operation is. Rather that it is an **extra** step to take, that occurs unpredictably.  
This is what affects real-time performance. You calculate/measure the timings of the application, and then randomly they may be violated.

Another thing is that `syslog_device` knows nothing about files. It may log to a file, or maybe to a character device.  
The only place that we know for sure that the log is on a file, and that it needs rotation is in `syslog_file_channel`.  
So how can `syslog_device` perform the rotation?